sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs

```diff
@@ -0,0 +1,412 @@
+            {
+                await plaintext.ReadAsync(new byte[gap], 0, gap).ConfigureAwait(false);
+            }
+
+            return new LengthLimitingStream(plaintext, encryptedBlobRange.OriginalRange.Count);
+        }
+
+        #endregion
+
+        private static Metadata ExtractMetadata(ResponseHeaders headers)
+        {
+            Metadata metadata = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+            foreach (HttpHeader header in headers)
+            {
+                if (header.Name.StartsWith(Constants.HeaderNames.MetadataPrefix, StringComparison.InvariantCulture))
+                {
+                    metadata[header.Name.Substring(Constants.HeaderNames.MetadataPrefix.Length)] = header.Value;
+                }
+            }
+
+            return metadata;
+        }
+
+        /// <summary>
+        /// Gets and validates a blob's encryption-related metadata
+        /// </summary>
+        /// <param name="metadata">The blob's metadata</param>
+        /// <returns>The relevant metadata.</returns>
+        private static EncryptionData? GetAndValidateEncryptionData(Metadata metadata)
+        {
+            _ = metadata ?? throw new InvalidOperationException();
+
+            EncryptionData encryptionData;
+            if (!metadata.TryGetValue(EncryptionConstants.ENCRYPTION_DATA_KEY, out string encryptedDataString))
+            {
+                return default;
+            }
+
+            using (var stream = new MemoryStream(Encoding.UTF8.GetBytes(encryptedDataString)))
+            {
+                var serializer = new DataContractJsonSerializer(typeof(EncryptionData));
+                encryptionData = (EncryptionData)serializer.ReadObject(stream);
+            }
+
+            _ = encryptionData.ContentEncryptionIV ?? throw new NullReferenceException();
+            _ = encryptionData.WrappedContentKey.EncryptedKey ?? throw new NullReferenceException();
+
+            // Throw if the encryption protocol on the message doesn't match the version that this client library
+            // understands and is able to decrypt.
+            if (EncryptionConstants.ENCRYPTION_PROTOCOL_V1 != encryptionData.EncryptionAgent.Protocol) {
+                throw EncryptionErrors.BadEncryptionAgent(encryptionData.EncryptionAgent.Protocol);
+            }
+
+            return encryptionData;
+        }
+
+        #region GetContentEncryptionKey
+
+        /// <summary>
+        /// Returns the key encryption key for blob. First tries to get key encryption key from KeyResolver, then
+        /// falls back to IKey stored on this EncryptionPolicy.
+        /// </summary>
+        /// <param name="encryptionData">The encryption data.</param>
+        /// <returns>Encryption key as a byte array.</returns>
+        private Memory<byte> GetContentEncryptionKey(EncryptionData encryptionData)
```

Comment:
I'm not going to put the comment on every method, but anything that's got sync/async pairs and isn't public should be merged into a single implementation.  Happy to walk through that pattern for the first few on Teams together if that's helpful.